vscode/src/debugger/session.ts

```diff
@@ -12,36 +12,85 @@ import {
   LoggingDebugSession,
   TerminatedEvent,
   logger,
+  Breakpoint,
+  StoppedEvent,
+  BreakpointEvent,
+  Thread,
+  StackFrame,
+  Source,
 } from "@vscode/debugadapter";
 
 import { FileAccessor } from "../common";
 import { DebugProtocol } from "@vscode/debugprotocol";
-import { ICompiler, log } from "qsharp";
+import {
+  BreakpointSpan,
+  ICompiler,
+  IDebugServiceWorker,
+  log,
+  Span,
+} from "qsharp";
 import { createDebugConsoleEventTarget } from "./output";
 import { ILaunchRequestArguments } from "./types";
 
 const ErrorProgramHasErrors = "program contains compile errors(s): cannot run.";
 const SimulationCompleted = "Q# simulation completed.";
 const ConfigurationDelayMS = 1000;
-const LaunchDelayMS = 100;
 
 function delay(ms: number): Promise<void> {
   return new Promise((resolve) => setTimeout(resolve, ms));
 }
 
 export class QscDebugSession extends LoggingDebugSession {
+  private static threadID = 1;
   private _fileAccessor: FileAccessor;
   private _compiler: ICompiler;
-
-  public constructor(fileAccessor: FileAccessor, compiler: ICompiler) {
+  private _debugger: IDebugServiceWorker;
+  private _config: vscode.DebugConfiguration;
+  private _breakpointLocations: Map<string, BreakpointSpan[]>;
+  private _breakpoints: Map<string, DebugProtocol.Breakpoint[]>;
+  private _failed: boolean;
+
+  public constructor(
+    fileAccessor: FileAccessor,
+    compiler: ICompiler,
+    debugService: IDebugServiceWorker,
+    config: vscode.DebugConfiguration
+  ) {
     super();
     this._fileAccessor = fileAccessor;
     this._compiler = compiler;
-
+    this._debugger = debugService;
+    this._config = config;
+    this._failed = false;
+    this._breakpointLocations = new Map<string, BreakpointSpan[]>();
+    this._breakpoints = new Map<string, DebugProtocol.Breakpoint[]>();
     this.setDebuggerLinesStartAt1(false);
     this.setDebuggerColumnsStartAt1(false);
   }
 
+  public async init(): Promise<void> {
+    const programText = await this._fileAccessor.readFileAsString(
+      this._config.program
+    );
+
+    const loaded = await this._debugger.loadSource(
+      this._config.program.path,
+      programText
+    );
+    if (loaded) {
+      const locations = await this._debugger.getBreakpoints(
+        this._config.program.path
+      );
+      log.trace(
+        `init breakpointLocations: ${JSON.stringify(locations, null, 2)}`
```

Comment:
You can use format specifiers (`%o`) to avoid JSON-ifying the string when tracing isn't enabled https://www.educative.io/answers/how-to-format-messages-in-consolelog-in-javascript  (this is what we do elsewhere in the code).